### ISSUE TITLE
Find/replace dialog fights me on selection

### ISSUE THREAD

**Issue by rgov** [microsoft/vscode#91210]
Issue Type: <b>Bug</b>

I find the find/replace dialog frustrating to use, in that it often switches modes on me against my wishes. I especially find that many of my frustrations stem from its behavior regarding the current document selection.

Let's suppose I have the following code:

```python
foo()

def asdf():
    foo()
```

I want to find the occurrence of `foo()` within the function body, ignoring the first line.

First I select the body of the function and press Cmd-F and type "foo".

Simple, right? What happened? Well, actually I haven't given you enough information to tell! It depends on whether I dragged from the top-left to select or the bottom-right, because this decides whether the insertion cursor is at the beginning of the function or after it.

If the insertion cursor was *before* the function definition, then the `foo` inside the function body is now highlighted. If the insertion cursor was *after* the function definition, then the topmost `foo` is highlighted, even though it was outside of my initial selection.

Either way, I lost my intended selection for scoping my search to just this function.

But, the selection is only hijacked when you modify the search query. If the dialog remembered the query from before, it worked as I expected. Before I carefully analyzed the behavior, this was one reason the dialog seemed unpredictable.

---

Now suppose I have closed the find dialog but later want to do the same search again. I select some code and again press Cmd-F, the dialog reopens. What state is it in?

Well, again the answer depends on context.

If I selected multiple lines of code, then my original search for "foo" is still in the field. But the "find in selection" option is reverted to being disabled.

If I selected a single line of code, or some portion of it, then the selected line is copied into the search field, erasing my previous search. If it was not "foo" but a complex regular expression, you can imagine this is frustrating.

---

In short, the find/replace dialog would benefit from more consistency of its behavior. Also, some of these behaviors may be expected on Windows but are not expected on macOS.

VS Code version: Code 1.42.1 (c47d83b293181d9be64f27ff093689e8e7aed054, 2020-02-11T14:44:27.652Z)
OS version: Darwin x64 19.3.0

**Comment by gonzojive** [comment]
I could not agree more with all the issues mentioned here. When I select a large block and hit Ctrl-F or Ctrl-H, I expect to be finding and replacing within the selection---why else would I bother selecting first?

The cryptic icons are even worse (issue #91211).

Considering how common copy/paste and find and replace is in today's programming world, this feature desperately needs tweaking.

**Comment by AlistairNorman** [comment]
Thanks for opening this issue. It's driving me nuts. I have to make a lot of small changes and the unexpected behaviour, in my case cursor jumping to the start of the searched term when I put it at the end of the searched term is pretty inconvenient.

**Comment by abulka** [comment]
Seems the secret to finding text in a selection is:
- ensure no text is selected
- invoke the `find` command (from the Edit menu or via running the command. The selection button toggle will be off.
- make your selection 
- toggle the selection button on

now you can type in your `Find` text and search within the selection.

Similarly, the secret to replacing text in a selection is:
- ensure no text is selected
- invoke the `replace` command (from the Edit menu or via running the command. The selection button toggle will be off.
- make your selection 
- toggle the selection button on

now you can type in your `Find` text and `Replace` text and things will work as needed within the selection .

Took me an hour of experimentation to finally figure this out. And that's after a year of not being quite sure and fighting vscode whenever I've needed to do it (and thinking it was buggy). Yes, its not obvious but at least now I know how to drive this vscode feature. Hope this helps someone else struggling with this.

**Comment by amunger** [comment]
"find in selection" is no longer automatically set when starting the find action, and I didn't see any of the other behavior when testing this on the latest version so I'll close this one

**Comment by gonzojive** [comment]
> "find in selection" is no longer automatically set when starting the find action, and I didn't see any of the other behavior when testing this on the latest version so I'll close this one

@amunger: Should I file a new feature request for the following?

> When I select a large block and hit Ctrl-F or Ctrl-H, I expect to be finding and replacing within the selection---why else would I bother selecting first?

**Comment by amunger** [comment]
you can set `"editor.find.autoFindInSelection": "multiline"` to enable that functionality.

**Comment by abulka** [comment]
OMG where has `"editor.find.autoFindInSelection": "multiline"` been all my life!
IMO this should be the default.

**Comment by rgov** [comment]
Does it include the whole line even if you only select part of it?

If you change selection and hit the hot key again does it update the search area?

**Comment by abulka** [comment]
> Does it include the whole line even if you only select part of it?

No.

> If you change selection and hit the hot key again does it update the search area?

No. You need to hit ESC to get out of the replace mode and then hit the hot key again - then the search area is updated (assuming you have changed the selection).
